### PR TITLE
:sparkles: Add variadic `apply` for `tuple`

### PR DESCRIPTION
**Pull request by elbeno**

### CODE REVIEW

File: test/tuple_algorithms.cpp [hunk 1]
```diff
@@ -90,6 +90,25 @@ TEST_CASE("apply", "[tuple_algorithms]") {
...
+
+    static_assert(stdx::apply([](auto x, auto y) { return x.value + y.value; },
+                              stdx::tuple{move_only{42}},
+                              stdx::tuple{move_only{17}}) == 59);
+
+    auto t = stdx::tuple{1, 2, 3};
+    stdx::apply([](auto &...xs) { (++xs, ...); }, t, t);
+    CHECK(t == stdx::tuple{3, 4, 5});
+}
+
```
[the remark this answers is not in the document]
Comment (lukevalenty) [reply test/tuple_algorithms.cpp:111]: Something like:

```c++
static_assert(stdx::apply([](auto... xs) { return stdx::make_tuple(xs...); }, 
                              stdx::tuple{1, 2, 3},
                              stdx::tuple{4, 5, 6}) == stdx::make_tuple(1, 2, 3, 4, 5, 6));
```

File: include/stdx/tuple_algorithms.hpp [hunk 2]
```diff
@@ -16,6 +16,29 @@
 
 namespace stdx {
 inline namespace v1 {
+template <typename F, tuplelike... Ts> constexpr auto apply(F &&f, Ts &&...ts) {
```
Comment (lukevalenty) [inline include/stdx/tuple_algorithms.hpp:19]: question: would it be worthwhile to write this in terms of `tuple_cat` or vis-a-versa?

File: test/tuple_algorithms.cpp [hunk 3]
```diff
@@ -90,6 +90,25 @@ TEST_CASE("apply", "[tuple_algorithms]") {
...
+    CHECK(stdx::apply(stateful) == 1);
+    CHECK(stdx::apply(stateful) == 2);
+
+    static_assert(stdx::apply([](auto x, auto y) { return x.value + y.value; },
+                              stdx::tuple{move_only{42}},
+                              stdx::tuple{move_only{17}}) == 59);
+
+    auto t = stdx::tuple{1, 2, 3};
+    stdx::apply([](auto &...xs) { (++xs, ...); }, t, t);
+    CHECK(t == stdx::tuple{3, 4, 5});
```
Comment (lukevalenty) [inline test/tuple_algorithms.cpp:109]: question: this is a really confusing test cause...took me several minutes to figure out what's going on. is there any way to make it more clear what it's intending to do? is it just checking that references to elements are getting passed correctly?

File: test/tuple_algorithms.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (elbeno) [reply test/tuple_algorithms.cpp]: Good call - added a couple of ordering tests.

File: test/tuple_algorithms.cpp (hunk 3 above)
Comment (elbeno) [reply test/tuple_algorithms.cpp]: Good call - improved it.

File: include/stdx/tuple_algorithms.hpp (hunk 2 above)
Comment (elbeno) [reply include/stdx/tuple_algorithms.hpp:19]: And the reason I didn't DRY `element_indices` is that I didn't want to adversely affect the compile-time performance of `tuple_cat`. It's a workhorse for CIB. Where variadic `apply` is something that I currently have one use case for (`when_all`).

**Review by lukevalenty — APPROVED** [review]

**Pull request merged**